src/zarr/abc/store.py

```diff
@@ -1,31 +1,72 @@
 from abc import ABC, abstractmethod
 from collections.abc import AsyncGenerator
-from typing import Protocol, runtime_checkable
+from typing import Any, NamedTuple, Protocol, runtime_checkable
+
+from typing_extensions import Self
 
 from zarr.buffer import Buffer, BufferPrototype
-from zarr.common import BytesLike, OpenMode
+from zarr.common import AccessModeLiteral, BytesLike
+
+
+class AccessMode(NamedTuple):
+    readonly: bool
+    overwrite: bool
+    create: bool
+    update: bool
+
+    @classmethod
+    def from_literal(cls, mode: AccessModeLiteral) -> Self:
+        if mode in ("r", "r+", "a", "w", "w-"):
+            return cls(
+                readonly=mode == "r",
+                overwrite=mode == "w",
+                create=mode in ("a", "w", "w-"),
+                update=mode in ("r+", "a"),
+            )
+        raise ValueError("mode must be one of 'r', 'r+', 'w', 'w-', 'a'")
 
 
 class Store(ABC):
-    _mode: OpenMode
+    _mode: AccessMode
+    _is_open: bool
+
+    def __init__(self, mode: AccessModeLiteral = "r", *args: Any, **kwargs: Any):
+        self._is_open = False
+        self._mode = AccessMode.from_literal(mode)
+
+    @classmethod
+    async def open(cls, *args: Any, **kwargs: Any) -> Self:
+        store = cls(*args, **kwargs)
+        await store._open()
+        return store
+
+    async def _open(self) -> None:
+        if await self.root_exists():
+            if self.mode.update or self.mode.readonly:
+                pass
+            elif self.mode.overwrite:
+                await self.clear()
+            else:
+                raise FileExistsError("Store already exists")
+        self._is_open = True
+
+    async def ensure_open(self) -> None:
+        if not self._is_open:
+            await self._open()
 
-    def __init__(self, mode: OpenMode = "r"):
-        if mode not in ("r", "r+", "w", "w-", "a"):
-            raise ValueError("mode must be one of 'r', 'r+', 'w', 'w-', 'a'")
-        self._mode = mode
+    @abstractmethod
+    async def root_exists(self) -> bool: ...
+
+    @abstractmethod
+    async def clear(self) -> None: ...
```

Comment:
let's add tests for both of these in `testing/store.py`

Reply:
good point! I added tests and changed `root_exists` to `not empty`